src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs

```diff
@@ -403,6 +403,23 @@ public virtual bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType
             return CheckTypesCore(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
         }
 
+        /// <summary>
+        /// Perform sub-expression type checking and produce a return type for the function declaration, this is only applicable for UDFs.
+        /// </summary>
+        public virtual bool CheckTypesOnDeclaration(CheckTypesContext context, TexlNode[] args, DType[] argTypes, DType returnType, IErrorContainer errors, out DType actualReturnType)
+        {
+            actualReturnType = DType.Unknown;
+
+            return false;
+        }
+
+        /// <summary>
+        /// Perform expression-level semantics checks which require a binding, this is only applicable for UDFs.
+        /// </summary>
+        public virtual void CheckSemanticsOnDeclaration(TexlBinding binding, TexlNode[] args, DType[] argTypes, IErrorContainer errors)
```

Comment:
Where are these new functions being called from?

Reply:
This question was never answered. Something needs to be calling these functions from within PowerFx to perform validation.

Reply:
Sorry, thought it's not applicable anymore. These function will be called from PAClient